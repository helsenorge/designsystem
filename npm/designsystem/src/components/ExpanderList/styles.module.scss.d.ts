export type Styles = {
  'expander-list': string;
  'expander-list__item': string;
  'expander-list__item--no-bottom': string;
  'expander-list__item--top': string;
  'expander-list-link': string;
  'expander-list-link__icon': string;
  'expander-list-link__icon-container--hasmargin': string;
  'expander-list-link__link-content': string;
  'expander-list-link__main-content': string;
  'expander-list-link__main-content--padding': string;
  'expander-list-link--banana': string;
  'expander-list-link--blueberry': string;
  'expander-list-link--cherry': string;
  'expander-list-link--closed': string;
  'expander-list-link--kiwi': string;
  'expander-list-link--large': string;
  'expander-list-link--neutral': string;
  'expander-list-link--plum': string;
  'expander-list-link--white': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
