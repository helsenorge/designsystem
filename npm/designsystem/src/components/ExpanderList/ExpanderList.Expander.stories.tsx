import React from 'react';

import { StoryObj, Meta } from '@storybook/react-vite';
import { Docs } from 'frankenstein-build-tools';

import ExpanderList, { Expander } from './ExpanderList';
import { allTitleTags } from '../../../.storybook/knobs';

const meta = {
  title: '@helsenorge/designsystem-react/Components/ExpanderList/ExpanderList.Expander',
  component: Expander,
  parameters: {
    docs: {
      page: (): React.JSX.Element => <Docs component={Expander} />,
      description: {
        component:
          'ExpanderList er en liste av elementer som skjuler detaljinformasjon når den ikke trengs, og gjør den lett tilgjengelig i kontekst når den trengs.',
      },
    },
  },
  args: {
    title: 'Her skrives en kort overskrift.',
    children: <span>{'Innhold i expander'}</span>,
  },
  argTypes: {
    id: {
      control: 'text',
    },
    title: {
      control: 'text',
    },
    titleHtmlMarkup: {
      control: 'select',
      options: allTitleTags,
      description: 'Changes the underlying element of the title. Default: span',
    },
    children: {
      control: 'object',
    },
    className: {
      control: 'text',
    },
    icon: {
      control: 'object',
    },
    expanded: {
      control: 'boolean',
      description: 'Controls the expander state',
    },
    testId: {
      control: 'text',
    },
    handleExpanderClick: { action: 'handleExpanderClick', description: 'Called when clicking the Expander.' },
    onExpand: { action: 'onExpand', description: 'Called when expander is open/closed.' },
    zIndex: {
      control: 'number',
      description: 'Overrides the default z-index of the expander header',
    },
    highlightText: {
      control: 'text',
      description: 'Highlights text in title and content. Override if different from list.',
    },
    status: {
      control: 'select',
      options: ['none', 'new'],
      description: 'Displays a status on the left side: default none',
    },
  },
} satisfies Meta<typeof Expander>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => (
    <>
      <ExpanderList>
        <ExpanderList.Expander {...args}>{args.children}</ExpanderList.Expander>
        <ExpanderList.Expander {...args}>{args.children}</ExpanderList.Expander>
      </ExpanderList>
    </>
  ),
};
