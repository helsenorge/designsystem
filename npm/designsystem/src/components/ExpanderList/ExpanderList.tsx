import React, { useEffect, useState } from 'react';

import { PaletteNames } from '../../theme/palette';
import Icon, { IconSize } from '../Icons';
import ChevronUp from '../Icons/ChevronUp';
import ChevronDown from '../Icons/ChevronDown';
import { useHover } from '../../hooks/useHover';
import { Breakpoint, useBreakpoint } from '../../hooks/useBreakpoint';
import { isElementInViewport } from '../../utils/viewport';

import classNames from 'classnames';

import expanderListStyles from './styles.module.scss';

export type ExpanderListColors = PaletteNames;
export type ExpanderType = React.ForwardRefExoticComponent<ExpanderProps & React.RefAttributes<HTMLLIElement>>;

export interface ExpanderListCompound extends React.ForwardRefExoticComponent<ExpanderListProps & React.RefAttributes<HTMLUListElement>> {
  Expander: ExpanderType;
}

interface MouseEventWithPath extends MouseEvent {
  path?: Array<HTMLElement>;
}

interface ExpanderListProps {
  /** Toggles accordion functionality for the expanders. */
  accordion?: boolean;
  /** Toggles the bottom border of the last child element. */
  bottomBorder?: boolean;
  /** Items in the ExpanderList */
  children: React.ReactNode;
  /** Toggles padding of child elements */
  childPadding?: boolean;
  /** Adds custom classes to the element. */
  className?: string;
  /** Changes the link list background color on hover. */
  color?: ExpanderListColors;
  /** Changes the font size. */
  large?: boolean;
  /** Opens the first item in the list. */
  isOpen?: boolean;
  /** Toggles the top border of the first child element. */
  topBorder?: boolean;
}

type Modify<T, R> = Omit<T, keyof R> & R;

type ExpanderProps = Modify<
  React.HTMLAttributes<HTMLButtonElement>,
  {
    id?: string;
    title: JSX.Element | string;
    children: React.ReactNode;
    className?: string;
    color?: ExpanderListColors;
    icon?: React.ReactElement;
    padding?: boolean;
    isExpanded?: boolean;
    large?: boolean;
    testId?: string;
    handleExpanderClick?: (event: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  }
>;

// TODO: See what can be done with regards to double reffing.
const Expander: ExpanderType = React.forwardRef((props: ExpanderProps, ref: React.Ref<HTMLLIElement>) => {
  const {
    id,
    children,
    padding = true,
    color = 'neutral',
    className = '',
    icon,
    large = false,
    title,
    isExpanded = false,
    testId = '',
    handleExpanderClick,
  } = props;
  const { hoverRef, isHovered } = useHover<HTMLButtonElement>();
  const breakpoint = useBreakpoint();
  const expanderClasses = classNames(expanderListStyles['expander-list-link'], expanderListStyles[`expander-list-link--${color}`], {
    [expanderListStyles['expander-list-link--closed']]: !isExpanded,
    [expanderListStyles['expander-list-link--large']]: large,
  });
  const mainContentClasses = classNames(
    expanderListStyles['expander-list-link__main-content'],
    padding ? expanderListStyles['expander-list-link__main-content--padding'] : ''
  );
  const iconContainerClasses = classNames(
    expanderListStyles['expander-list-link__icon-container'],
    expanderListStyles['expander-list-link__icon-container--hasmargin']
  );

  return (
    <li className={className} ref={ref}>
      <button id={id} onClick={handleExpanderClick} data-testid={testId} className={expanderClasses} ref={hoverRef}>
        <span className={expanderListStyles['expander-list-link__link-content']}>
          {icon ? (
            <span className={iconContainerClasses}>
              {React.cloneElement(icon, {
                className: expanderListStyles['expander-list-link__icon'],
                size: breakpoint === Breakpoint.Xs ? IconSize.XSmall : IconSize.Small,
                isHovered,
              })}
            </span>
          ) : null}
          {title}
        </span>
        <span className={expanderListStyles['expander-list-link__icon-container']}>
          <Icon
            className={expanderListStyles['expander-list-link__icon']}
            size={IconSize.XSmall}
            svgIcon={isExpanded ? ChevronUp : ChevronDown}
            isHovered={isHovered}
          />
        </span>
      </button>
      {isExpanded ? <div className={mainContentClasses}>{children}</div> : null}
    </li>
  );
});

function findShadowDOMId(event: MouseEventWithPath, tagName: string): string {
  const pathElements: Array<HTMLElement> = event.path || (event.composedPath && (event.composedPath() as Array<HTMLElement>));
  const element = pathElements.find((el: HTMLElement) => el.tagName === tagName);
  return element ? element.id : '';
}

export const ExpanderList = React.forwardRef((props: ExpanderListProps, ref: React.Ref<HTMLUListElement>) => {
  const {
    children,
    childPadding = true,
    large,
    isOpen = false,
    color,
    className = '',
    accordion = false,
    topBorder = true,
    bottomBorder = true,
  } = props;
  const [activeExpander, setActiveExpander] = useState({});
  const [latestExpander, setLatestExpander] = useState<HTMLElement>();
  const childCount = React.Children.count(children);
  const expanderListClasses = classNames(expanderListStyles['expander-list'], className);

  function handleExpanderClick(event: React.MouseEvent<HTMLElement, MouseEvent>): void {
    const id = event.currentTarget?.id || findShadowDOMId((event as unknown) as MouseEventWithPath, 'BUTTON');
    if (!isOpen) {
      setActiveExpander(prevState => (accordion ? { [id]: !prevState[id] } : { ...prevState, [id]: !prevState[id] }));
      setLatestExpander(event.currentTarget);
    }
  }

  /** Returns the class modifier top when we want to show the top border and no-bottom when we don't want to show the bottom border */
  const getExpanderItemClass = (index: number) => {
    return classNames(expanderListStyles['expander-list__item'], {
      [expanderListStyles['expander-list__item--top']]: index === 0 && topBorder,
      [expanderListStyles['expander-list__item--no-bottom']]: index === childCount - 1 && !bottomBorder,
    });
  };

  useEffect(() => {
    if (accordion && latestExpander && !isElementInViewport(latestExpander)) {
      latestExpander.scrollIntoView();
    }
  }, [accordion, latestExpander]);

  return (
    <ul className={expanderListClasses} ref={ref}>
      {React.Children.map(children, (child: React.ReactNode, index: number) => {
        if ((child as React.ReactElement<ExpanderProps>).type === Expander) {
          const isExpanded = isOpen || activeExpander[`expander-${index}`];
          const expanderItemClass = getExpanderItemClass(index);

          return React.cloneElement(child as React.ReactElement<ExpanderProps>, {
            id: `expander-${index}`,
            key: `expander-${index}`,
            isExpanded,
            padding: childPadding,
            color,
            large,
            'aria-expanded': isExpanded,
            className: expanderItemClass,
            handleExpanderClick: handleExpanderClick,
          });
        }
        return child;
      })}
    </ul>
  );
}) as ExpanderListCompound;

ExpanderList.Expander = Expander;
Expander.displayName = 'ExpanderList.Expander';

export default ExpanderList;
