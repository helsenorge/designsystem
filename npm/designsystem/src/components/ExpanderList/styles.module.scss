@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_font-settings.scss';
@import '../../scss/_palette.scss';

.expander-list {
  list-style: none;
  padding: 0;
  width: inherit;
  margin: 0;

  &__item {
    border: 0;
    border-bottom: 1px solid $neutral500;

    &--top {
      border-top: 1px solid $neutral500;
    }

    &--no-bottom {
      border-bottom: 0;
    }
  }
}

.expander-list-link {
  padding: getSpacer(xs) 0 getSpacer(xs) 0;
  font-size: $font-size-sm;
  line-height: $lineheight-size-sm;
  @media (min-width: map-get($grid-breakpoints, md)) {
    padding-left: getSpacer(xs);
    padding-right: getSpacer(xs);
    font-size: $font-size-md;
    line-height: $lineheight-size-md;
  }
  display: flex;
  cursor: pointer;
  font-weight: 600;
  font-family: inherit;
  text-decoration: none;
  justify-content: space-between;
  align-items: center;
  letter-spacing: unset;
  color: $black;
  border: 0;
  width: 100%;
  outline: none;

  &--banana {
    background-color: $banana100;

    &:hover,
    &:focus {
      background-color: $banana100;
    }
  }
  &--blueberry {
    background-color: $blueberry100;

    &:hover,
    &:focus {
      background-color: $blueberry100;
    }
  }
  &--cherry {
    background-color: $cherry100;

    &:hover,
    &:focus {
      background-color: $cherry100;
    }
  }
  &--kiwi {
    background-color: $kiwi100;

    &:hover,
    &:focus {
      background-color: $kiwi100;
    }
  }
  &--neutral {
    background-color: $neutral100;

    &:hover,
    &:focus {
      background-color: $neutral100;
    }
  }
  &--plum {
    background-color: $plum100;

    &:hover,
    &:focus {
      background-color: $plum100;
    }
  }
  &--white {
    background-color: $neutral50;

    &:hover,
    &:focus {
      background-color: $neutral50;
    }
  }

  &--closed:not(:hover, :focus) {
    background-color: transparent;
  }

  &:focus {
    outline: none;
  }

  &--large {
    font-size: $font-size-lg;
    line-height: $lineheight-size-lg;
  }

  &__link-content {
    color: $black;
    display: flex;
    align-items: center;
    text-align: left;
    border-bottom: 1px solid transparent;

    :focus > & {
      border-color: $black;
    }
  }

  &__main-content {
    padding: 0;

    &--padding {
      padding: getSpacer(m) 0 getSpacer(m) 0;
      @media (min-width: map-get($grid-breakpoints, md)) {
        padding-left: getSpacer(xs);
        padding-right: getSpacer(xs);
      }
    }
  }

  &__icon-container {
    &--hasmargin {
      margin-right: getSpacer(xs);
    }
  }
  &__icon {
    vertical-align: middle;
  }
}
