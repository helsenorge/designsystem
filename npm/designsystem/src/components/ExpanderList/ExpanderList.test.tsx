import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ExpanderList from './ExpanderList';
import * as ViewportUtils from '../../utils/viewport';
import Icon from '../Icon';
import Avatar from '../Icons/Avatar';
import Title from '../Title/Title';
import '../../__mocks__/uuid';

describe('Gitt ExpanderList blir rendret', (): void => {
  describe('Når det er tre expandere og accordion er false', (): void => {
    it('Sjekk at ExpanderList sitt snapshot matcher', (): void => {
      const { container } = render(
        <ExpanderList accordion={false} childPadding={true} color={'blueberry'}>
          <ExpanderList.Expander title="Title 1">{'Text 1'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Når det er to expandere og den første er expanded', (): void => {
    it('så er den første expanderen åpen, og den andre ikke åpen', async (): Promise<void> => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title="Title 1" expanded>
            {'Text 1'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button1 = screen.getByRole('button', { name: 'Title 1' });
      expect(button1).toHaveAttribute('aria-expanded', 'true');

      const button2 = screen.getByRole('button', { name: 'Title 2' });
      expect(button2).toHaveAttribute('aria-expanded', 'false');
    });

    it('så kan man åpne den andre expanderen', async (): Promise<void> => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title="Title 1" expanded>
            {'Text 1'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button2 = screen.getByRole('button', { name: 'Title 2' });
      expect(button2).toHaveAttribute('aria-expanded', 'false');

      await userEvent.click(button2);

      expect(button2).toHaveAttribute('aria-expanded', 'true');
      const text2 = screen.getByText('Text 2');
      expect(text2).toBeInTheDocument();
    });

    it('så kan man lukke den første', async (): Promise<void> => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title="Title 1" expanded>
            {'Text 1'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button1 = screen.getByRole('button', { name: 'Title 1' });
      expect(button1).toHaveAttribute('aria-expanded', 'true');

      const text1 = screen.getByText('Text 1');
      expect(text1).toBeInTheDocument();

      await userEvent.click(button1);

      expect(button1).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når det er to expandere med expanded true', (): void => {
    it('så er begge expanderne åpne', async (): Promise<void> => {
      render(
        <ExpanderList>
          <ExpanderList.Expander expanded title="Title 1">
            {'Text 1'}
          </ExpanderList.Expander>
          <ExpanderList.Expander expanded title="Title 2">
            {'Text 2'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 3'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button1 = screen.getByRole('button', { name: 'Title 1' });
      expect(button1).toHaveAttribute('aria-expanded', 'true');

      const button2 = screen.getByRole('button', { name: 'Title 2' });
      expect(button2).toHaveAttribute('aria-expanded', 'true');

      const button3 = screen.getByRole('button', { name: 'Title 3' });
      expect(button3).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når det er to expandere, der den andre har expanded true', (): void => {
    it('så er bare den andre expanderen åpen', async (): Promise<void> => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title="Title 1">{'Text 1'}</ExpanderList.Expander>
          <ExpanderList.Expander expanded title="Title 2">
            {'Text 2'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 3'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button1 = screen.getByRole('button', { name: 'Title 1' });
      expect(button1).toHaveAttribute('aria-expanded', 'false');

      const button2 = screen.getByRole('button', { name: 'Title 2' });
      expect(button2).toHaveAttribute('aria-expanded', 'true');

      const button3 = screen.getByRole('button', { name: 'Title 3' });
      expect(button3).toHaveAttribute('aria-expanded', 'false');
    });
  });

  describe('Når rerendring trigges i children mens expander er åpen', (): void => {
    it('så er expander fortsatt åpen', async (): Promise<void> => {
      const ExpanderListTest: React.FC = () => {
        const [text, setText] = React.useState('knapp');

        return (
          <ExpanderList>
            <ExpanderList.Expander title={'Title 1'}>
              <button onClick={() => setText('oppdatert knapp')}>{text}</button>
            </ExpanderList.Expander>
          </ExpanderList>
        );
      };

      render(<ExpanderListTest />);

      const expander = screen.getByRole('button', { name: 'Title 1' });
      await userEvent.click(expander);
      expect(expander).toHaveAttribute('aria-expanded', 'true');

      const button = screen.getByRole('button', { name: 'knapp' });
      await userEvent.click(button);

      expect(expander).toHaveAttribute('aria-expanded', 'true');
    });
  });

  describe('Når det er en expander der children kan oppdatere state', (): void => {
    describe('Når state oppdateres i children mens expander er åpen', (): void => {
      it('så er expander fortsatt åpen', async (): Promise<void> => {
        const ExpanderListTest: React.FC = () => {
          const [text, setText] = React.useState('knapp');

          return (
            <ExpanderList>
              <ExpanderList.Expander title={'Title 1'}>
                <button onClick={() => setText('oppdatert knapp')}>{text}</button>
              </ExpanderList.Expander>
            </ExpanderList>
          );
        };

        render(<ExpanderListTest />);

        const expander = screen.getByRole('button', { name: 'Title 1' });
        await userEvent.click(expander);
        expect(expander).toHaveAttribute('aria-expanded', 'true');

        const button = screen.getByRole('button', { name: 'knapp' });
        await userEvent.click(button);

        expect(expander).toHaveAttribute('aria-expanded', 'true');
      });
    });
  });

  describe('Når det er en expander der expanded-status kan endres utenfra', (): void => {
    describe('Når expanded endrer status', (): void => {
      it('Så toggles riktig expander', async (): Promise<void> => {
        const ExpanderListTest: React.FC = () => {
          const [isExpanded, setIsExpanded] = React.useState(false);

          return (
            <>
              <ExpanderList>
                <ExpanderList.Expander title={'Title 1'}>{'Hei 1'}</ExpanderList.Expander>
                <ExpanderList.Expander title={'Title 2'} expanded={isExpanded}>
                  {'Hei 2'}
                </ExpanderList.Expander>
              </ExpanderList>
              <button onClick={() => setIsExpanded(!isExpanded)}>{isExpanded ? 'Lukk' : 'Åpne'}</button>
            </>
          );
        };

        render(<ExpanderListTest />);

        const expander1 = screen.getByRole('button', { name: 'Title 1' });
        expect(expander1).toHaveAttribute('aria-expanded', 'false');

        const expander2 = screen.getByRole('button', { name: 'Title 2' });
        expect(expander2).toHaveAttribute('aria-expanded', 'false');

        await userEvent.click(screen.getByRole('button', { name: 'Åpne' }));

        expect(expander1).toHaveAttribute('aria-expanded', 'false');
        expect(expander2).toHaveAttribute('aria-expanded', 'true');

        await userEvent.click(screen.getByRole('button', { name: 'Lukk' }));
        expect(expander1).toHaveAttribute('aria-expanded', 'false');
        expect(expander2).toHaveAttribute('aria-expanded', 'false');
      });
    });
  });

  describe('Når man klikker på en expander', (): void => {
    it('Sjekk Expander bakgrunnsfarge når musepeker trykker på den', async (): Promise<void> => {
      render(
        <div
          style={{
            width: '40rem',
          }}
        >
          <ExpanderList accordion={false} childPadding={true} color={'blueberry'}>
            <ExpanderList.Expander testId="test01" title="Title 1">
              {'Text 1'}
            </ExpanderList.Expander>
            <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
            <ExpanderList.Expander title="Title 3">{'Text 2'}</ExpanderList.Expander>
          </ExpanderList>
        </div>
      );

      const button = screen.getByRole('button', { name: 'Title 1' });
      await userEvent.click(button);
      expect(button).toHaveClass('expander-list-link expander-list-link--blueberry');
    });
  });

  describe('Når accordion er true', (): void => {
    test('Så vises bare en tekst om gangen', async (): Promise<void> => {
      render(
        <ExpanderList accordion={true}>
          <ExpanderList.Expander title="Title 1">{'Text 1'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 3'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const button1 = screen.getByRole('button', { name: 'Title 1' });
      await userEvent.click(button1);

      expect(button1).toHaveAttribute('aria-expanded', 'true');

      const text1 = screen.getByText('Text 1');

      expect(text1).toBeInTheDocument();

      const button2 = screen.getByRole('button', { name: 'Title 2' });
      await userEvent.click(button2);

      expect(button1).toHaveAttribute('aria-expanded', 'false');
      expect(button2).toHaveAttribute('aria-expanded', 'true');

      const text2 = screen.getByText('Text 2');

      expect(text2).toBeInTheDocument();
    });
  });

  describe('Når accordion er true og expanderen er utenfor viewport etter at man klikker på den', (): void => {
    test('Så scroller nettleseren slik at expanderen blir synlig', async (): Promise<void> => {
      const mockScrollIntoView = vi.fn();
      window.HTMLElement.prototype.scrollIntoView = mockScrollIntoView;
      const mockIsElementInViewport = vi.spyOn(ViewportUtils, 'isElementInViewport');

      mockIsElementInViewport.mockReturnValue(false);
      render(
        <ExpanderList accordion={true}>
          <ExpanderList.Expander title="Title 1">{'Text 1'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const title1 = screen.getByRole('button', { name: 'Title 3' });
      await userEvent.click(title1);

      expect(mockScrollIntoView).toHaveBeenCalledTimes(1);
      expect(mockScrollIntoView).toHaveBeenCalledWith();
    });
  });

  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(
        <ExpanderList testId="bare-tester">
          <ExpanderList.Expander title="Title 1">{'Text 1'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 2">{'Text 2'}</ExpanderList.Expander>
          <ExpanderList.Expander title="Title 3">{'Text 2'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });
  describe('Gitt at ExpanderList vises', (): void => {
    test('Så har den riktig analyticsid', (): void => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title="Title 1" testId="expander">
            {'Text 1'}
          </ExpanderList.Expander>
        </ExpanderList>
      );

      const button = screen.getByTestId('expander');

      expect(button).toHaveAttribute('data-analyticsid', 'expander-list-expander');
    });
  });
  describe('Gitt at ExpanderList har expander med onExpand-callback', (): void => {
    describe('Når man klikker på ekspanderen tre ganger', (): void => {
      test('Så kalles callbacken med true, false og true', async (): Promise<void> => {
        const handleExpand = vi.fn();
        render(
          <ExpanderList>
            <ExpanderList.Expander title="Title 1" onExpand={handleExpand}>
              {'Text 1'}
            </ExpanderList.Expander>
          </ExpanderList>
        );

        const expander = screen.getByText('Title 1');

        await userEvent.click(expander);
        await userEvent.click(expander);
        await userEvent.click(expander);
        expect(handleExpand).toHaveBeenCalledTimes(3);
        expect(handleExpand).toHaveBeenNthCalledWith(1, true);
        expect(handleExpand).toHaveBeenNthCalledWith(2, false);
        expect(handleExpand).toHaveBeenNthCalledWith(3, true);
      });
    });
  });
  describe('Gitt at ExpanderList har Expander med onExpand-callback og den første er expanded', (): void => {
    describe('Når man klikker på expanderen én gang', (): void => {
      test('Så kalles callback først med true og så med false', async (): Promise<void> => {
        const handleExpand = vi.fn();
        render(
          <ExpanderList>
            <ExpanderList.Expander title="Title 1" onExpand={handleExpand} expanded>
              {'Text 1'}
            </ExpanderList.Expander>
          </ExpanderList>
        );

        const expander = screen.getByText('Title 1');

        await userEvent.click(expander);
        expect(handleExpand).toHaveBeenCalledTimes(2);
        expect(handleExpand).toHaveBeenNthCalledWith(1, true);
        expect(handleExpand).toHaveBeenNthCalledWith(2, false);
      });
    });
  });
  describe('Når det er to ExpanderList i samme dokument', (): void => {
    test('Så har expanderne forskjellige IDer', (): void => {
      const { container } = render(
        <>
          <ExpanderList>
            <ExpanderList.Expander title="ExpanderList 1">{'Text 1'}</ExpanderList.Expander>
          </ExpanderList>
          <ExpanderList>
            <ExpanderList.Expander title="ExpanderList 2">{'Text 2'}</ExpanderList.Expander>
          </ExpanderList>
        </>
      );

      const buttonId1 = screen.getByRole('button', { name: 'ExpanderList 1' }).id;
      const buttonId2 = screen.getByRole('button', { name: 'ExpanderList 2' }).id;

      expect(buttonId1).not.toEqual(buttonId2);
      expect(container).toMatchSnapshot();
    });
  });
  describe('Når title er JSX', (): void => {
    test('Så fungerer expanderen', (): void => {
      const { container } = render(
        <ExpanderList>
          <ExpanderList.Expander
            titleHtmlMarkup="span"
            title={
              <span style={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between', width: '100%' }}>
                <Title appearance="title3">{'ExpanderList 1'}</Title>
                <Icon svgIcon={Avatar} />
              </span>
            }
          >
            {'Hei'}
          </ExpanderList.Expander>
          <ExpanderList.Expander title={<Title appearance="title3">{'ExpanderList 2'}</Title>}>{'Hei'}</ExpanderList.Expander>
        </ExpanderList>
      );

      const buttonId1 = screen.getByRole('button', { name: 'ExpanderList 1' }).id;
      const buttonId2 = screen.getByRole('button', { name: 'ExpanderList 2' }).id;

      expect(buttonId1).not.toEqual(buttonId2);
      expect(container).toMatchSnapshot();
    });
  });
  describe('Når renderChildrenWhenClosed ikke er satt', (): void => {
    test('Så er children ikke rendret', (): void => {
      render(
        <ExpanderList>
          <ExpanderList.Expander title={'Test'}>
            <span data-testid="test">{'Test'}</span>
          </ExpanderList.Expander>
        </ExpanderList>
      );

      const child = screen.queryByTestId('test');
      expect(child).not.toBeInTheDocument();
    });
  });
  describe('Når renderChildrenWhenClosed er true', (): void => {
    test('Så er children rendret', (): void => {
      render(
        <ExpanderList renderChildrenWhenClosed={true}>
          <ExpanderList.Expander title={'Test'}>
            <span data-testid="test">{'Test'}</span>
          </ExpanderList.Expander>
        </ExpanderList>
      );

      const child = screen.getByTestId('test');
      expect(child).toBeInTheDocument();
    });
  });
  describe('Når renderChildrenWhenClosed er false', (): void => {
    test('Så er children ikke rendret', (): void => {
      render(
        <ExpanderList renderChildrenWhenClosed={false}>
          <ExpanderList.Expander title={'Test'}>
            <span data-testid="test">{'Test'}</span>
          </ExpanderList.Expander>
        </ExpanderList>
      );

      const child = screen.queryByTestId('test');
      expect(child).not.toBeInTheDocument();
    });
  });
  describe('Når title er en react-komponent uten children', (): void => {
    test('Så vises tittelen', (): void => {
      const TestTitle: React.FC = () => (
        <div>
          <h1>{'Dette er en tittel'}</h1>
        </div>
      );

      render(
        <ExpanderList>
          <ExpanderList.Expander title={<TestTitle />}>
            <span data-testid="test">{'Test'}</span>
          </ExpanderList.Expander>
        </ExpanderList>
      );

      const title = screen.getByRole('heading', { name: 'Dette er en tittel' });
      expect(title).toBeVisible();
    });
  });
});
