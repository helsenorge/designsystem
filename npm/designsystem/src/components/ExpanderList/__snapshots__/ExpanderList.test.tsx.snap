// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt ExpanderList blir rendret NÃ¥r det er tre expandere og accordion er false Sjekk at ExpanderList sitt snapshot matcher 1`] = `
<div>
  <div
    style="width: 40rem;"
  >
    <ul
      class="expander-list"
    >
      <li
        class="expander-list__item expander-list__item--top"
      >
        <button
          class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
          data-testid=""
          id="expander-0"
        >
          <span
            class="expander-list-link__link-content"
          >
            Title 1
          </span>
          <span
            class="expander-list-link__icon-container"
          >
            <svg
              aria-hidden="true"
              class="hnds-style-icon expander-list-link__icon"
              fill="black"
              height="38"
              role="img"
              style="min-width: 38px; min-height: 38px;"
              viewBox="0 0 48 48"
              width="38"
            >
              <path
                d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
              />
            </svg>
          </span>
        </button>
      </li>
      <li
        class="expander-list__item"
      >
        <button
          class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
          data-testid=""
          id="expander-1"
        >
          <span
            class="expander-list-link__link-content"
          >
            Title 2
          </span>
          <span
            class="expander-list-link__icon-container"
          >
            <svg
              aria-hidden="true"
              class="hnds-style-icon expander-list-link__icon"
              fill="black"
              height="38"
              role="img"
              style="min-width: 38px; min-height: 38px;"
              viewBox="0 0 48 48"
              width="38"
            >
              <path
                d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
              />
            </svg>
          </span>
        </button>
      </li>
      <li
        class="expander-list__item"
      >
        <button
          class="expander-list-link expander-list-link--blueberry expander-list-link--closed"
          data-testid=""
          id="expander-2"
        >
          <span
            class="expander-list-link__link-content"
          >
            Title 3
          </span>
          <span
            class="expander-list-link__icon-container"
          >
            <svg
              aria-hidden="true"
              class="hnds-style-icon expander-list-link__icon"
              fill="black"
              height="38"
              role="img"
              style="min-width: 38px; min-height: 38px;"
              viewBox="0 0 48 48"
              width="38"
            >
              <path
                d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
              />
            </svg>
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
